test/14.errorHandling.js

```diff
@@ -0,0 +1,412 @@
+    const update_lock = true;
+    const post_commit = true;
+
+    await c1.execute("set autocommit off");
+    await c2.execute("set autocommit off");
+
+    if (update_lock) {
+      await c2.execute("update T2 set F1=F1+1");
+      await c1.execute("update T1 set F1=F1+1");
+    }
+
+    // now produce a deadlock condition
+    let err = null;
+    try {
+      await Promise.all([
+        c1.execute("update T2 set F1=F1+1"),
+        c2.execute("update T1 set F1=F1+1"),
+      ]);
+    } catch (e) {
+      err = e;
+      console.error(e);
+    }
+
+    if (post_commit) {
+      await c1.commit();
+      await c2.commit();
+    }
+
+    should.exist(err);
+
+    // close the connections
+    await c1.close();
+    await c2.close();
+  });
+
+  it("14.2 Can detect an index uniqueness error", async () => {
+    let err = null;
+    try {
+      await conn.execute("INSERT INTO T3 VALUES (1)");
+      await conn.execute("INSERT INTO T3 VALUES (1)");
+    } catch (e) {
+      err = e;
+      console.error(e);
+    }
+    should.exist(err);
+  });
+
+  it("14.3 Can detect when a TE goes down before query execution", async () => {
+    const newTE = await startTE(JSON.stringify(postData));
+    let err = null;
+    // wait to do anything until the TE is running.
+    let TEProcData;
+    try {
+      TEProcData = await pollForTERunning(newTE.startId)
+    } catch (e) {
+      should.not.exist(e);
+    }
+
+    // create a connection to this TE
+    const directConnection = await driver.connect({
+      ...config,
+      LBQuery: `round_robin(start_id(${newTE.startId}))`,
+    });
+
+    // kill the TE
+    try {
+      await killTE(TEProcData.pid);
+    } catch (e) {
+      should.not.exist(e);
+    }
+
+    // try to execute a query on the TE
+    let results;
+    try {
+      results = await directConnection.execute(
+        "select getnodeid() from dual"
+      );
+      await results.getRows();
+      await results.close();
+    } catch (e) {
+      err = e;
+      console.error(e);
+    }
+    should.exist(err);
+
+    try {
+      await results.close();
+      await directConnection.close();
+    } catch (e) {
+      // do nothing
+    }
+  });
+
+  it("14.4 Can detect when a TE goes down after query execution", async () => {
+    const newTE = await startTE(JSON.stringify(postData));
+    let err = null;
+    // wait to do anything until the TE is running.
+    let TEProcData;
+    try {
+      TEProcData = await pollForTERunning(newTE.startId)
+    } catch (e) {
+      should.not.exist(e);
+    }
+
+    // create a connection to this TE
+    const directConnection = await driver.connect({
+      ...config,
+      LBQuery: `round_robin(start_id(${newTE.startId}))`,
+    });
+
+    // try to execute a query on the TE, kill before getting results
+    let results;
+    try {
+      results = await directConnection.execute(
+        "select getnodeid() from dual"
+      );
+      await killTE(TEProcData.pid);
+      await results.getRows();
```

Comment:
I presume this always works and there is never the chance of getting a first buffer read before it has to hit the server, but I do not know the exact implementation of how the driver manages row retrieval, the model I have in my head is just an abstraction.  It may be worth double checking with the SQL guys, but if you saw this work while testing we are probably OK.

Reply:
The driver does not request the rows until getRows is called. It's managed in the NAN layer. The ResultSet object is just an abstraction of a reference to a C Driver ResultSetObject